lib/auth.js

```diff
@@ -0,0 +1,248 @@
+        const [localPart, domain] = login.split('@');
+        return domain.length <= MAX_DOMAIN_LENGTH &&
+          localPart.length <= MAX_LOCAL_PART_LENGTH &&
+          EMAIL_REGEXP.test(login);
+      }
+    },
+    hint: () => 'Login must be email address'
+  }
+};
+
+const loginPasswordTests = {
+  LOGIN_INCLUDES_PASSWORD: {
+    test: (login, password) => !login.includes(password),
+    hint: () => 'The login can not contain a password'
+  },
+  PASSWORD_INCLUDES_LOGIN: {
+    test: (login, password) => !password.includes(login),
+    hint: () => 'The password can not contain a login'
+  }
+};
+
+class AuthenticationStrength {
+  constructor(
+    valid, // Boolean
+    hints, // Object, { required: Array, optional: Array }
+    compliance // Number, percent of best test group
+  ) {
+    this.valid = valid;
+    this.hints = hints;
+    this.strength = !valid ? 'Not valid' :
+      compliance < 0.25 ? 'Very weak' :
+        compliance < 0.4 ? 'Weak' :
+          compliance < 0.6 ? 'Good' :
+            compliance < 0.8 ? 'Strong' : 'Very strong';
+  }
+}
+
+const makeTest = (
+  // Function that checks the arguments on a test suite
+  tests, // Array, array of password/login tests
+  required, // Array, required tests configs
+  optional, // Array, optional tests configs
+  ...testArgs // Array, [password] / [login] / [login, password]
+  // Returns: AuthenticationStrength instance
+) => {
+  const test = testsConfig => {
+    const testsHints = [];
+    testsConfig.forEach(testConfig => {
+      const [testName, userOptions] = typeof testConfig === 'string' ?
+        [testConfig, {}] : [testConfig.name, testConfig];
+      const { test, hint, options = {} } = tests[testName];
+      const testOptions = Object.assign({}, options, userOptions);
+      if (!test(...testArgs, testOptions)) testsHints.push(hint(testOptions));
+    });
+    return testsHints;
+  };
+
+  const requiredHints = test(required);
+  const optionalHints = test(optional);
+
+  const valid = !requiredHints.length;
+  const compliance = 1 - (optional.length ?
+    optionalHints.length / optional.length : 0);
+  const hints = { required: requiredHints, optional: optionalHints };
+
+  return new AuthenticationStrength(valid, hints, compliance);
+};
+
+const checkLogin = (
+  // Function that tests the login
+  login, // String, login to test
+  required, // Array, required tests configs
+  optional = [] // Array, optional tests configs
+  // Returns: AuthenticationStrength instance
+) => {
+  if (!required) {
+    required = ['MIN_LENGTH', 'MAX_LENGTH'];
+  }
+  return makeTest(loginTests, required, optional, login);
+};
+
+const checkPassword = (
+  // Function that tests the password
+  password, // String, password to test
+  required, // Array, required tests configs
+  optional = [] // Array, optional tests configs
+  // Returns: AuthenticationStrength instance
+) => {
+  if (!required) {
+    required = ['MIN_LENGTH', 'MAX_LENGTH'];
+    optional = [
```

Comment:
Why do you reassign optional tests if the user doesn't provide the required tests?
What if the user wants to provide the optional tests and use the default required tests?

Reply:
Now there are two options: the user uses the default configuration and passes only the test data to the function, or the user uses its configuration.
Now the user can not pass optional tests without passing the required ones, so if the user wants to use any of his own configurations, he must pass full test configuration.